// @stackbit/sdk/src/config/presets-loader.ts
import _ from 'lodash';
import path from 'path';
import fse from 'fs-extra';
import { parseFile, append } from '@stackbit/utils';

import { Config, Preset } from './config-types';
import { ConfigPresetsError } from './config-errors';

export interface PresetsLoaderResult {
    config: Config;
    errors: ConfigPresetsError[];
}

interface RawPresetData {
    model: string;
    presets: RawPreset[];
}

interface RawPreset {
    label: string;
    thumbnail?: string;
    data: Record<string, any>;
}

export async function loadPresets(dirPath: string, config: Config): Promise<PresetsLoaderResult> {
    const presetFiles = [];
    let presetsRelDirs = ['.stackbit/presets', 'node_modules/@stackbit/components/presets'];

    if (config.presetSource?.type === 'files' && config.presetSource?.presetDirs) {
        presetsRelDirs = [...new Set([...presetsRelDirs, ...config.presetSource.presetDirs])];
    }

    for (const presetsRelDir of presetsRelDirs) {
        const presetsDir = path.join(dirPath, presetsRelDir);
        if (!(await fse.pathExists(presetsDir))) {
            continue;
        }
        const files = (await fse.readdir(presetsDir))
            .filter((fileName) => ['.json', '.yaml', '.yml'].includes(path.parse(fileName).ext))
            .map((fileName) => path.join(presetsRelDir, fileName));
        presetFiles.push(...files);
    }

    const presets: Record<string, Preset> = {};
    const presetsIdsByModel: Record<string, string[]> = {};
    const errors: ConfigPresetsError[] = [];

    for (const presetFile of presetFiles) {
        const presetsRelDir = path.dirname(presetFile);
        const presetPath = path.join(dirPath, presetFile);
        let presetData: RawPresetData;
        try {
            presetData = await parseFile(presetPath);
        } catch (err: any) {
            errors.push(new ConfigPresetsError(`Error parsing ${presetFile} (${err?.message})`));
            continue;
        }
        _.forEach(_.get(presetData, 'presets', []), (preset, i) => {
            const presetId = `${presetFile}:presets[${i}]`;
            const { thumbnail, ...rest } = preset;
            presets[presetId] = {
                ...rest,
                ...(thumbnail ? { thumbnail: resolveThumbnailPath(thumbnail, presetsRelDir) } : null),
                modelName: presetData.model
            };
            append(presetsIdsByModel, presetData.model, presetId);
        });
    }

    // update models with presets IDs
    const models = _.map(config.models, (model) => {
        const presetIdsForModel = presetsIdsByModel[model.name];
        if (!presetIdsForModel) {
            return model;
        }
        return { ...model, presets: presetIdsForModel };
    });

    return {
        config: Object.assign({}, config, {
            models,
            presets
        }),
        errors
    };
}

function resolveThumbnailPath(thumbnail: string, dir: string) {
    if (thumbnail.startsWith('/')) {
        if (dir.endsWith('@stackbit/components/presets')) {
            dir = dir.replace(/\/presets$/, '');
        } else {
            dir = '';
        }
        thumbnail = thumbnail.replace(/^\//, '');
    }
    return path.join(dir, thumbnail);
}
